From: RoccoDev <roccodev.business@gmail.com>
Date: Tue, 15 Sep 2020 11:49:12 +0200
Subject: [PATCH] Make cacti physics break trigger BlockPhysicsEvent

Specifically, cacti breaking due to another block being next to them.

diff --git a/src/main/java/net/minecraft/server/BlockCactus.java b/src/main/java/net/minecraft/server/BlockCactus.java
index ddc2135..3bb1a5f 100644
--- a/src/main/java/net/minecraft/server/BlockCactus.java
+++ b/src/main/java/net/minecraft/server/BlockCactus.java
@@ -1,10 +1,12 @@
 package net.minecraft.server;
 
+import org.bukkit.craftbukkit.event.CraftEventFactory;
+import org.bukkit.craftbukkit.util.CraftMagicNumbers;
+import org.bukkit.event.block.BlockPhysicsEvent;
+
 import java.util.Iterator;
 import java.util.Random;
 
-import org.bukkit.craftbukkit.event.CraftEventFactory; // CraftBukkit
-
 public class BlockCactus extends Block {
 
     public static final BlockStateInteger AGE = BlockStateInteger.of("age", 0, 15);
@@ -64,9 +66,11 @@ public class BlockCactus extends Block {
 
     public void doPhysics(World world, BlockPosition blockposition, IBlockData iblockdata, Block block) {
         if (!this.e(world, blockposition)) {
-            world.setAir(blockposition, true);
+            org.bukkit.block.Block bukkit = world.getWorld().getBlockAt(blockposition.getX(), blockposition.getY(), blockposition.getZ());
+            BlockPhysicsEvent event = new BlockPhysicsEvent(bukkit, bukkit.getTypeId());
+            world.getServer().getPluginManager().callEvent(event);
+            if(!event.isCancelled()) world.setAir(blockposition, true);
         }
-
     }
 
     public boolean e(World world, BlockPosition blockposition) {
-- 
2.48.1

